src/transcode/interface.ts

```diff
@@ -0,0 +1,21 @@
+import { EventFragment, Interface, Result } from 'ethers/lib/utils';
+
+export class WarpInterface {
+  ethersInterface: Interface;
+  decodeEventLog(eventFragment: EventFragment | string, data: string, topics: string[]): Result {
+    return this.ethersInterface.decodeEventLog(eventFragment, data, topics);
+  }
+  encodeEventLog(
+    eventFragment: EventFragment | string,
+    values: any[],
+  ): { data: string; topics: Array<string> } {
+    return this.ethersInterface.encodeEventLog(eventFragment, values);
```

Comment:
This is a neat way, I think we can get an even better experience by:
- making WarpInterface inherit from Interface 
- putting the logic of `encodeWarpEvent` and `decodeWarpEvent` in `encodeEventLog` and `decodeEventLog`

By doing libraries programmed against ethers.js API can easily use our replacement without changing any API.

Reply:
But the problem is we can't override the extended class function if types/numbers of arguments are different.

Reply:
decodeWarpEvent has the same signature and changing the signature of encodeWarpEvent shouldn't be too hard

Reply:
decodeWarpEvent has signature `(fragment: EventFragment | string, warpEvent: EventItem)` while decodeEventLog has signature `(eventFragment: EventFragment | string, data: string, topics: string[])`

Reply:
Problem is data array we receive is of type `string[]` while decodeEventLog expects data as `BytesLikes` so we can't completely put all logic of `warpDecodeEvent` inside `decodeEventLog` keeping the type of data same

Reply:
what do you think about the following re-structuring:
```ts
export class WarpInterface extends Interface {
  decodeWarpEvent(fragment: EventFragment | string, warpEvent: EventItem): Result {
    warpEvent = join248bitChunks(warpEvent); // reverse 248 bit packing

    // Remove leading 0x from each element and join them
    const data = `0x${warpEvent.data.map((x) => x.slice(2)).join('')}`;

    const result = super.decodeEventLog(fragment, data, warpEvent.keys);
    return result;
  }

  encodeWarpEvent(fragment: EventFragment, values: argType[], order = 0): EventItem {
    const { data, topics }: { data: string; topics: string[] } = super.encodeEventLog(
      fragment,
      values,
    );

    const topicFlatHex = '0x' + topics.map((x) => x.slice(2).padStart(64, '0')).join('');
    const topicItems248: string[] = splitInto248BitChunks(topicFlatHex);
    const dataItems248: string[] = splitInto248BitChunks(data);

    return { order, keys: topicItems248, data: dataItems248 };
  }
}

```

Reply:
Yea this looks good